File: services/fastify-service/src/routes/health.ts
import { FastifyInstance, FastifyPluginCallback } from "fastify";

let isReady = false;

export function setReady(ready: boolean): void {
  isReady = ready;
}

const healthRoutes: FastifyPluginCallback = (
  fastify: FastifyInstance,
  _opts,
  done
) => {
  fastify.get("/health", async () => {
    return { status: "ok", timestamp: new Date().toISOString() };
  });

  fastify.get("/ready", async (_request, reply) => {
    if (isReady) {
      return { status: "ready", timestamp: new Date().toISOString() };
    }
    reply.status(503);
    return { status: "not_ready", timestamp: new Date().toISOString() };
  });

  done();
};

export default healthRoutes;

File: services/fastapi-service/app/routes/health.py
from datetime import datetime, timezone

from fastapi import APIRouter

router = APIRouter()

_is_ready = False


def set_ready(ready: bool):
    global _is_ready
    _is_ready = ready


@router.get("/health")
async def health():
    return {"status": "ok", "timestamp": datetime.now(timezone.utc).isoformat()}


@router.get("/ready")
async def ready():
    if _is_ready:
        return {"status": "ready", "timestamp": datetime.now(timezone.utc).isoformat()}
    from fastapi.responses import JSONResponse

    return JSONResponse(
        status_code=503,
        content={"status": "not_ready", "timestamp": datetime.now(timezone.utc).isoformat()},
    )
